### PR TITLE
Bug 654

### PR DESCRIPTION
**Pull request by rzezeski**
A proposed patch for bug-654.

I created the generic function extract_headers so that adding other custom headers down the line (for example, to set the tree order) would be simple.  E.g. just add {tree_order, ?HEAD_TREE_ORDER} to the list and it's good to go.

I have no logical reason for choosing a custom header vs. a query param...it just _felt_ better to me.  I'd be happy to add or change this to use a query param instead.

### CODE REVIEW

**Comment by beerriot** [comment]
The header value needs a little more frobbing: luwak_file expects block_size to be an integer, but this code sets it as a string.

**Comment by rzezeski** [comment]
On my first patch I simply did a curl and checked the file object, but didn't verify the contents.  This time I was a little more thorough; I went so far as to generate the graph returned by visualize_tree.

File: src/luwak_wm_file.erl [hunk 1]
```diff
@@ -543,6 +545,24 @@ extract_user_meta(RD) ->
...
+                    []).
+
+extract_headers(_RD, [], Acc) ->
+    Acc;
+extract_headers(RD, [{Key, Header, Cast}|T], Acc) ->
+    case wrq:get_req_header(Header, RD) of
+        undefined ->
+            extract_headers(RD, T, Acc);
+        Val ->
+            extract_headers(RD, T, [{Key, Cast(Val)}|Acc])
```
Comment (Vagabond) [inline src/luwak_wm_file.erl]: Maybe do the cast in a try/catch so you can gracefully return errors if the block size isn't an integer. Right now it blows up pretty dramatically.

File: src/luwak_wm_file.erl [hunk 2]
```diff
@@ -495,7 +496,8 @@ accept_doc_body(RD, Ctx=#ctx{key=K, client=C}) ->
     H0 = case Ctx#ctx.handle of
              {ok, H} -> H;
              _ ->
-                 {ok, H} = luwak_file:create(C, K, dict:new()),
+                 FileProps = extract_file_props(RD),
+                 {ok, H} = luwak_file:create(C, K, FileProps, dict:new()),
```
Comment (Vagabond) [inline src/luwak_wm_file.erl]: This wasn't a concern before, but now that you can pass in block-size, luwak_file:create will throw() if the block size is less than 1. You should probably catch that and convert it to a decent error message.

**Comment by rzezeski** [comment]
I moved the header parsing to `malformed_request` where it belongs and added validation.  I think the `malformed_request` bodies could be cleaned up by using `error_resp` but that's out of scope for this pull request.

**Comment by Vagabond** [comment]
Merged.

**Pull request closed without merging**